build_tools/lint.fish

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env fish
+#
+# This is meant to be run by "make lint" or "make lint-all". It is not meant to
+# be run directly from a shell prompt.
+#
+set cppchecks warning,performance,portability,information,missingInclude
+set cppcheck_args
+set c_files
+set all no
+
+set -gx CXX $argv[1]
+set -e argv[1]
+
+if test "$argv[1]" = "--all"
+    set all yes
+    set c_files src/
+    set cppchecks "$cppchecks,unusedFunction"
+    set -e argv[1]
+end
+
+# We only want -D and -I options to be passed thru to cppcheck.
+for arg in $argv
+    if string match -q -- '-D*' $arg
+        set cppcheck_args $cppcheck_args $arg
+    else if string match -q -- '-I*' $arg
+        set cppcheck_args $cppcheck_args $arg
+    end
+end
+if test (uname -m) = "x86_64"
+    set cppcheck_args -D__x86_64__ -D__LP64__ $cppcheck_args
+end
+
+if test $all = no
+    # We haven't been asked to lint all the source. If there are uncommitted
+    # changes lint those, else lint the files in the most recent commit.
+    set pending (git status --short --untracked-files=all | sed -e 's/^ *//')
```

Comment:
Yep, it's to strip the leading whitespace from lines like this: ` M CONTRIBUTING.md`. Without trimming that whitespace the string split below doesn't work as expected. I thought about string trim but that seemed less clear in this situation. Thanks for the suggestion to add `--porcelain`. Done.